File: packages/core/src/agent-store/scanner.ts
import { readdir, lstat, readlink } from "node:fs/promises";
import { join } from "node:path";
import { fileExists } from "../utils/fs.js";
import type { AgentType, AgentPathConfig, ProjectAgentScanResult } from "./types.js";
import { AGENT_PATHS } from "./types.js";

/**
 * Scan a single project directory for agent configurations.
 * Detects .claude/ and .gemini/ directories and catalogs their contents.
 */
export async function scanProject(
  projectName: string,
  projectPath: string,
): Promise<ProjectAgentScanResult> {
  const result: ProjectAgentScanResult = {
    projectName,
    projectPath,
    agents: {},
  };

  await Promise.all(
    (Object.entries(AGENT_PATHS) as [AgentType, AgentPathConfig][]).map(
      async ([agent, paths]) => {
        const rootDir = join(projectPath, paths.root);
        if (!(await fileExists(rootDir))) return;

        const [skills, commands, hooks, hasMemoryFile, hasMcpConfig] =
          await Promise.all([
            listSubdirs(join(projectPath, paths.skills)),
            listMdFiles(join(projectPath, paths.commands)),
            listFiles(join(projectPath, paths.hooks)),
            fileExists(join(projectPath, paths.memoryFile)),
            fileExists(join(projectPath, paths.mcpConfig)),
          ]);

        result.agents[agent] = {
          hasConfig: true,
          skills,
          commands,
          hooks,
          hasMemoryFile,
          hasMcpConfig,
        };
      },
    ),
  );

  return result;
}

/**
 * Scan all projects in the workspace for agent configurations.
 */
export async function scanAllProjects(
  projects: Array<{ name: string; path: string }>,
  workspaceRoot: string,
): Promise<ProjectAgentScanResult[]> {
  return Promise.all(
    projects.map((p) => scanProject(p.name, join(workspaceRoot, p.path))),
  );
}

/**
 * Check if a path is a symlink and resolve its target.
 */
export async function checkSymlink(path: string): Promise<{
  isSymlink: boolean;
  target?: string;
}> {
  try {
    const stats = await lstat(path);
    if (stats.isSymbolicLink()) {
      const target = await readlink(path);
      return { isSymlink: true, target };
    }
    return { isSymlink: false };
  } catch {
    return { isSymlink: false };
  }
}

/** List subdirectory names (for skills — each skill is a folder) */
async function listSubdirs(dir: string): Promise<string[]> {
  try {
    const entries = await readdir(dir, { withFileTypes: true });
    return entries.filter((e) => e.isDirectory()).map((e) => e.name);
  } catch {
    return [];
  }
}

/** List .md filenames without extension (for commands) */
async function listMdFiles(dir: string): Promise<string[]> {
  try {
    const entries = await readdir(dir, { withFileTypes: true });
    return entries
      .filter((e) => e.isFile() && e.name.endsWith(".md"))
      .map((e) => e.name.slice(0, -3));
  } catch {
    return [];
  }
}

/** List all filenames (for hooks) */
async function listFiles(dir: string): Promise<string[]> {
  try {
    const entries = await readdir(dir, { withFileTypes: true });
    return entries.filter((e) => e.isFile()).map((e) => e.name);
  } catch {
    return [];
  }
}

File: server/src/agent_store/scanner.rs
use std::path::{Path, PathBuf};
use tokio::fs;

use super::schema::{AgentPresence, AgentType, ProjectAgentScanResult, agent_paths};

/// Scan a single project directory for agent configurations.
pub async fn scan_project(project_name: &str, project_path: &Path) -> ProjectAgentScanResult {
    let mut result = ProjectAgentScanResult {
        project_name: project_name.to_string(),
        project_path: project_path.to_path_buf(),
        agents: Default::default(),
    };

    let agents = [AgentType::Claude, AgentType::Gemini];

    for agent in agents {
        let paths = agent_paths(agent);
        let root_dir = project_path.join(paths.root);
        if !root_dir.exists() {
            continue;
        }

        let skills = list_subdirs(&project_path.join(paths.skills)).await;
        let commands = list_md_files(&project_path.join(paths.commands)).await;
        let hooks = list_files(&project_path.join(paths.hooks)).await;
        let has_memory_file = project_path.join(paths.memory_file).exists();
        let has_mcp_config = project_path.join(paths.mcp_config).exists();

        result.agents.insert(
            agent.to_string(),
            AgentPresence {
                has_config: true,
                skills,
                commands,
                hooks,
                has_memory_file,
                has_mcp_config,
            },
        );
    }

    result
}

/// Scan all projects in the workspace for agent configurations.
pub async fn scan_all_projects(
    projects: &[(&str, &Path)],
    workspace_root: &Path,
) -> Vec<ProjectAgentScanResult> {
    let mut results = Vec::with_capacity(projects.len());
    for (name, rel_path) in projects {
        let abs_path = workspace_root.join(rel_path);
        results.push(scan_project(name, &abs_path).await);
    }
    results
}

/// Check if a path is a symlink and resolve its target.
pub async fn check_symlink(path: &Path) -> (bool, Option<PathBuf>) {
    match fs::symlink_metadata(path).await {
        Ok(meta) if meta.file_type().is_symlink() => match fs::read_link(path).await {
            Ok(target) => (true, Some(target)),
            Err(_) => (true, None),
        },
        _ => (false, None),
    }
}

async fn list_subdirs(dir: &Path) -> Vec<String> {
    let mut result = Vec::new();
    let Ok(mut rd) = fs::read_dir(dir).await else {
        return result;
    };
    while let Ok(Some(entry)) = rd.next_entry().await {
        if let Ok(ft) = entry.file_type().await {
            if ft.is_dir() {
                result.push(entry.file_name().to_string_lossy().into_owned());
            }
        }
    }
    result
}

async fn list_md_files(dir: &Path) -> Vec<String> {
    let mut result = Vec::new();
    let Ok(mut rd) = fs::read_dir(dir).await else {
        return result;
    };
    while let Ok(Some(entry)) = rd.next_entry().await {
        let name = entry.file_name().to_string_lossy().into_owned();
        if name.ends_with(".md") {
            if let Ok(ft) = entry.file_type().await {
                if ft.is_file() {
                    result.push(name[..name.len() - 3].to_string());
                }
            }
        }
    }
    result
}

async fn list_files(dir: &Path) -> Vec<String> {
    let mut result = Vec::new();
    let Ok(mut rd) = fs::read_dir(dir).await else {
        return result;
    };
    while let Ok(Some(entry)) = rd.next_entry().await {
        if let Ok(ft) = entry.file_type().await {
            if ft.is_file() {
                result.push(entry.file_name().to_string_lossy().into_owned());
            }
        }
    }
    result
}
